### PR TITLE
[DebugInfo] Fix debug info for bitfield members

### PR DESCRIPTION
**Pull request by asudarsa**
Debug info for bitfield members was not translated. Adding it here.

Thanks

### CODE REVIEW

**Comment by asudarsa** [comment]
I will add a test for this in a bit. thanks

**Review by MrSidims — COMMENTED** [review]
Thanks! Lgtm, just waiting for the test

**Comment by MrSidims** [comment]
Okay, it seems like it's incorrect, since OpenCL.DebugInfo.100 defines flags ending up with 1 << 16 for FlagTypePassByReference . Moreover 1 << 17 is occupied by NonSemantic.DebugInfo.100 FlagUnknownPhysicalLayout . Guess, we need to redefine Flags table in NonSemantic.DebugInfo.200 may be adding there several more entries from ./include/llvm/IR/DebugInfoFlags.def , WDYT? Note, it would mean, the the bitfield flag would remain untranslatable until we switch to new version of DebugInfo

**Comment by asudarsa** [comment]
> Okay, thanks to the tests, it seems like it's incorrect, since OpenCL.DebugInfo.100 defines flags ending up with 1 << 16 for FlagTypePassByReference . Moreover 1 << 17 is occupied by NonSemantic.DebugInfo.100 FlagUnknownPhysicalLayout . Guess, we need to redefine Flags table in NonSemantic.DebugInfo.200 may be adding there several more entries from ./include/llvm/IR/DebugInfoFlags.def , WDYT? Note, it would mean, the the bitfield flag would remain untranslatable until we switch to new version of DebugInfo

Ah..good observation..thanks Dmitry. I am not sure if there is any other option. :-(. How long do you estimate for the switch?

Thanks

**Comment by MrSidims** [comment]
@svenvh I see that ARM added bitfield clang extension for OpenCL, would you find useful adding SPIRVDebug::FlagBitField to OpenCL.DebugInfo.100? If so we can commit this patch as is disabling validation tests and then update Registry and Tools to make this flag known.
Also tagging @bashbaug to see if it's legit suggestion.

**Comment by svenvh** [comment]
> @svenvh I see that ARM added bitfield clang extension for OpenCL, would you find useful adding SPIRVDebug::FlagBitField to OpenCL.DebugInfo.100? If so we can commit this patch as is (or moving it to 18'a bit) disabling validation tests and then update Registry and Tools to make this flag known. Also tagging @bashbaug to see if it's legit suggestion.

I suppose you're referring to the `__cl_clang_bitfields` extension?  It's not a portable extension, so it may not be worth spending a lot of effort in adding it to DebugInfo (but if the work has already been done, we could I suppose).

I haven't followed the DebugInfo work much to be honest so I may not be the best person to advise on this. :)

File: test/bitfields_packed.ll [hunk 1]
```diff
@@ -0,0 +1,124 @@
+; RUN: llvm-as %s -o %t.bc
+; RUN: llvm-spirv --spirv-debug-info-version=nonsemantic-shader-200 %t.bc -o %t.spv
+; RUN: llvm-spirv --spirv-debug-info-version=nonsemantic-shader-200 %t.spv -o %t.spt --to-text
+; RUN: llvm-spirv --spirv-debug-info-version=nonsemantic-shader-200 -r -emit-opaque-pointers %t.spv -o %t.bc
```
Comment (MrSidims) [inline test/bitfields_packed.ll:4]:
```suggestion
; RUN: llvm-spirv -r -emit-opaque-pointers %t.spv -o %t.bc
```

File: test/bitfields_packed.ll [hunk 2]
```diff
@@ -0,0 +1,124 @@
+; RUN: llvm-as %s -o %t.bc
+; RUN: llvm-spirv --spirv-debug-info-version=nonsemantic-shader-200 %t.bc -o %t.spv
+; RUN: llvm-spirv --spirv-debug-info-version=nonsemantic-shader-200 %t.spv -o %t.spt --to-text
```
Comment (MrSidims) [inline test/bitfields_packed.ll:3]:
```suggestion
; RUN: llvm-spirv %t.spv -o %t.spt --to-text
```

File: test/bitfields_packed.ll [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+; RUN: llvm-spirv --spirv-debug-info-version=nonsemantic-shader-200 -r -emit-opaque-pointers %t.spv -o %t.bc
+; RUN: llvm-dis %t.bc -o %t.ll
```
Comment (MrSidims) [inline test/bitfields_packed.ll:5]: Please also add
; RUN: llvm-spirv --spirv-debug-info-version=ocl-100
and
; RUN: llvm-spirv --spirv-debug-info-version=nonsemantic-shader-100

without CHECK lines - to check whether the test doesn't crash.

**Review by MrSidims — COMMENTED** [review]
LGTM, 1 test nit

File: test/bitfields_packed.ll [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+; RUN: FileCheck %s --input-file %t_1.ll  -check-prefix=CHECK-LLVM
+
+;CHECK-SPIRV: String [[#A:]] "a"
+;CHECK-SPIRV: String [[#B:]] "b"
+;CHECK-SPIRV: String [[#C:]] "c"
+;CHECK-SPIRV: String [[#D:]] "d"
+;CHECK-SPIRV: String [[#E:]] "e"
+;CHECK-SPIRV: String [[#F:]] "f"
+
+;CHECK-SPIRV: Constant [[#]] [[#]] 3
```
Comment (MrSidims) [inline test/bitfields_packed.ll:18]: we should also check of constant 262144
and following DebugTypeMember should contain it

File: test/bitfields_packed.ll [hunk 5]
```diff
@@ -0,0 +1,125 @@
+; RUN: llvm-as %s -o %t.bc
+; RUN: llvm-spirv --spirv-debug-info-version=ocl-100 %t.bc
+; RUN: llvm-spirv --spirv-ext=+SPV_KHR_non_semantic_info --spirv-debug-info-version=nonsemantic-shader-100 %t.bc
```
Comment (MrSidims) [inline test/bitfields_packed.ll:3]:
```suggestion
; RUN: llvm-spirv --spirv-debug-info-version=nonsemantic-shader-100 %t.bc
```

File: test/bitfields_packed.ll [hunk 6]
```diff
@@ -0,0 +1,125 @@
+; RUN: llvm-as %s -o %t.bc
```
Comment (MrSidims) [inline test/bitfields_packed.ll:1]: Please move the test to test/DebugInfo/NonSemantic/Shader200 dir

File: test/bitfields_packed.ll [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+;CHECK-SPIRV: String [[#E:]] "e"
+;CHECK-SPIRV: String [[#F:]] "f"
+
+;CHECK-SPIRV: Constant [[#]] [[#]] 3
+;CHECK-SPIRV: Constant [[#]] [[#THREE:]] 3
+;CHECK-SPIRV: Constant [[#]] [[#]] 1
+;CHECK-SPIRV: Constant [[#]] [[#]] 1
+;CHECK-SPIRV: Constant [[#]] [[#ONE:]] 1
+
+;CHECK-SPIRV: ExtInst [[#]] [[#]] [[#]] DebugTypeMember [[#A]] [[#]] [[#]] [[#]] [[#]] [[#]] [[#]] [[#THREE]]
```
Comment (MrSidims) [inline test/bitfields_packed.ll:24]: Okay, seems like we have something not implemented for DebugTypeMember (not related with your patch) - it no longer has Scope operand. Will be fixed later.

File: test/bitfields_packed.ll (hunk 5 above)
Comment (asudarsa) [reply test/bitfields_packed.ll:3]: I added this because local testing failed without this.

Comment (MrSidims) [reply test/bitfields_packed.ll:3]: Okay, I'll take a look at this later

File: test/bitfields_packed.ll (hunk 4 above)
Comment (asudarsa) [reply test/bitfields_packed.ll:18]: I do not see this in the .spt file. Is this related to the 'scope' issue you mentioned later? Thanks

**Review by MrSidims — APPROVED** [review]

**Comment by MrSidims** [comment]
out-of-tree is passing
in-tree will be fixed by a patch from Sven

**Pull request merged**